### ISSUE TITLE
"NotSupportedError: ReadableStream uploading is not supported" errors in Safari

### ISSUE THREAD

**Issue by juliandormon** [GoogleChrome/workbox#1732]
**Library Affected**:
workbox-sw

**Browser & Platform**:
Safari 12 on iOS 12

**Issue or Feature Request Description**:
When submitting a form that is registered with a route, I get the following error:
"NotSupportedError: ReadableStream uploading is not supported"

Everything works perfectly on Android and Chrome desktop on my Mac.

Here is the javascript, I'm using to register the routes:

```
const bgSyncPlugin = new workbox.backgroundSync.Plugin('ppSortMagicQueue');

    workbox.routing.registerRoute(
        '/Home/_SaveGallery',
        workbox.strategies.networkOnly({
            plugins: [bgSyncPlugin]
        }),
        'POST'
    );
```

**Here is the request info from the debugger Network tab:**

```
Summary
URL: https://sortmagic.redacted.net/Home/_SaveGallery
Status: —
Source: —

Request
Content-Type: application/x-www-form-urlencoded; charset=UTF-8
Accept: */*
Referer: https://sortmagic.redacted.net/
Origin: https://sortmagic.redacted.net
User-Agent: Mozilla/5.0 (iPad; CPU OS 12_0_1 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/12.0 Mobile/15E148 Safari/604.1
X-Requested-With: XMLHttpRequest

Response
No response headers

Request Data
MIME Type: application/x-www-form-urlencoded; charset=UTF-8
WebsiteId: 5140
EventID: 7671
EventName
token: Cl+ndF3l7owkH1+3bcO2AJQYB2gIfDtqt65uSDzoV2t8jeZZ+pf1ncOUn/1fntxv2lUS0T11AipCNgwUxiql7MGWyfDItQXY5tVsKYHWp/IByhrTJtZde1+E2t8IC+GE
GalleryName: jQKJ17Ij8
Password
X-Requested-With: XMLHttpRequest

**And here is the console output from Workbox:**

[Error] workbox – "Network request for '/Home/_SaveGallery' threw an error."
NotSupportedError: ReadableStream uploading is not supported
fetch
(anonymous function) — workbox-core.dev.js:1400
generatorResume
step — workbox-core.dev.js:14
(anonymous function) — workbox-core.dev.js:32
initializePromise
Promise
(anonymous function) — workbox-core.dev.js:11
(anonymous function) — workbox-strategies.dev.js:822
generatorResume
step — workbox-core.dev.js:14
(anonymous function) — workbox-core.dev.js:32
initializePromise
Promise
(anonymous function) — workbox-core.dev.js:11
generatorResume
step — workbox-core.dev.js:14
(anonymous function) — workbox-core.dev.js:32
initializePromise
Promise
(anonymous function) — workbox-core.dev.js:11
handleRequest — workbox-routing.dev.js:365
(anonymous function) — workbox-routing.dev.js:835
	_print (workbox-core.dev.js:132)
	(anonymous function) (workbox-core.dev.js:1407)
	generatorResume
	step (workbox-core.dev.js:14)
	(anonymous function) (workbox-core.dev.js:27)
	promiseReactionJob
```

Thanks for your help!

**Comment by philipwalton** [comment]
Hmmm, but this workaround assumes the requests all have text bodies, no? Currently we're constructing all replayed requests with `Blob` bodies to support binary files and whatnot.

**Comment by jeffposnick** [comment]
Oh—I hadn't looked at the code, and assumed, based on the error, that it was just using `request.body` directly as the `BodyInit`, but I didn't realize it was currently using the `blob()` reader:

https://github.com/GoogleChrome/workbox/blob/b146d8d3c2db89fa51afc322049b9795894bb4f2/packages/workbox-background-sync/models/StorableRequest.mjs#L52-L54

I guess I'm curious as to what Safari considers a non-`ReadableStream` upload, in that case? Should we ask some folks more familiar with the Streams API?

**Comment by juliandormon** [comment]
Thanks for checking on this guys. Is there something I can do to work around this from my end or do I need to wait for an update in your code?

**Comment by juliandormon** [comment]
Hey guys! Good news I figured out the issue. While testing on another iOS device, everything worked as expected. I wanted to send you the debug info and after I connected the iPhone to my laptop via USB and started the Safari debugging console, submitted the form, and the error occurred. After disconnecting, the post worked. I suspect the ReadableStream is used in the debugging process.

Please feel free to close this if you're satisfies with that solution. Thanks!

**Comment by jeffposnick** [comment]
I'm happy to close in that case. We can revisit if someone else can trigger this outside of the debugging scenario you describe.

**Comment by philipwalton** [comment]
So, I notice this error today as well when looking at the JS console in Safari. It seems to be happening as part of the `workbox-google-analytics` package, which is using a `NetworkFirst` strategy and for some reason throwing a `no-response` `WorkboxError`.

@jeffposnick, any ideas?

<img width="824" alt="screen shot 2018-11-09 at 9 57 48 pm" src="https://user-images.githubusercontent.com/326742/48298036-074e3680-e46b-11e8-9074-af1991c82b4c.png">

Note: this doesn't seem to be related to background sync, at least as far as I can tell.

**Comment by philipwalton** [comment]
Actually, it does appear like the problem is exactly as described in that Stack Overflow post. These are POST requests, so service worker in Safari doesn't appear to be able to replay them.

Since, all Google Analytics requests sent with `navigator.sendBeacon()` are all going to be POST requests, this seems like a relatively big problem...

**Comment by scurry84** [comment]
I'm having this same problem with Safari v12 on OSX.

**Comment by jeffposnick** [comment]
I dug into this a bit more and found the following, using https://glitch.com/edit/#!/suave-beanie as a simple reproduction.

Generally speaking, Safari has no problem sending a `POST` request constructed with a `body`.

The problem, which was accurately identified in [this comment on Stack Overflow](https://stackoverflow.com/questions/49850705/fetch-in-safari-11-1-request-with-console-log-doesnt-work-readablestream-upl/50952018#comment86715917_49850705), is that Safari's behavior changes based on whether or not the `Request` object is logged to the console before it's being sent. Apparently, whatever code is necessary to log it is enough to convert the underlying `Request` object into a `ReadableStream`, and that means it's no longer valid when sending the `POST`. Additionally, you apparently need to have to have the JS console open to trigger it—either the main console or the dedicated service worker console, depending on which context is doing the logging.

(This sounds like a bug that should be reported against WebKit, because I can't imagine that's ideal behavior.)

`workbox-routing` will log the `Request` object prior to passing it to `fetch()` at:

https://github.com/GoogleChrome/workbox/blob/194cdeb63d5abb21490f88f01f02f4bcf7e6d54b/packages/workbox-routing/Router.mjs#L200-L204

We should either remove that logging statement, or alternatively only log when it's `GET` request or when we know there's no `body`.

To summarize, this is a real bug that you can encounter in Safari for any  request with a `body` that is handled by `workbox-routing`, but only if
- you are [using the `dev` bundles of Workbox](https://developers.google.com/web/tools/workbox/guides/configure-workbox#configure_debug_builds_vs_production_builds) which triggers extra logging
- you have the Safari service worker debugger console open prior to the request is being sent

**Comment by edwardkenfox** [comment]
The previously mentioned [WebKit issue](https://bugs.webkit.org/show_bug.cgi?id=203617) has been addressed quite recently and states that

> Logging to the console isn't the problem, it's trying to access the `body` property at all.

which seems be contrary to the observation https://github.com/GoogleChrome/workbox/issues/1732#issuecomment-548001615 that 

> Additionally, you apparently need to have to have the JS console open to trigger it—either the main console or the dedicated service worker console, depending on which context is doing the logging

I understand this bug had already been fixed in workbox, but I've come across this behavior in my personal project and can't get my head around it so forgive me for bringing this up again...How was this behavior of whether having the JS console open or not changing the existence of the error observed? With my version of macOS Safari (Version 15.0) it seems like this is not the case and the error `ReadableStream uploading is not supported` appears even when the JS console is not opened (I also tried turning off the Developer options from Preferences). 

Would appreciate any input on this matter, thank you.
